src/magic_folder/magic_folder.py

```diff
@@ -857,3 +874,97 @@ def add_file(self, path):
         d = defer.Deferred()
         self._queue.put((bytespath, d))
         return d
+
+@attr.s
+class RemoteSnapshotCreator(object):
+    _state_db = attr.ib()
+    _local_author = attr.ib()
+
+    @eliotutil.inline_callbacks
+    def _upload_local_snapshots(self, tahoe_client):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+
+        :param tahoe_client: a TahoeClient instance
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._state_db.get_all_localsnapshot_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
+        for name in localsnapshot_names:
+            action = UPLOADER_SERVICE_UPLOAD_LOCAL_SNAPSHOTS(relpath=name)
+            with action:
+                # deserialize into LocalSnapshot
+                snapshot = self._state_db.get_local_snapshot(name, self._local_author)
+
+                # now upload each item in the queue
+                try:
+                    remote_snapshot = yield write_snapshot_to_tahoe(
+                        snapshot,
+                        self._local_author,
+                        tahoe_client,
+                    )
+
+                    # At this point, remote snapshot creation successful for
+                    # the given relpath.
+                    # store the remote snapshot capability in the db.
+                    yield self._state_db.store_remotesnapshot(name, remote_snapshot)
+
+                    # Remove the local snapshot content from the stash area.
+                    snapshot.content_path.remove()
+
+                    # Remove the LocalSnapshot from the db.
+                    yield self._state_db.delete_localsnapshot(name)
+
+
+                except NoServersError:
```

Comment:
How would this ever get thrown? We will use `tahoe_client.create_immutable_directory()` under the hood which can only throw HTTP errors, AFAICT.

Reply:
Oh, right. I was thinking that we would get a bubbled up "connection failed" socket error. However that is not the case since that error would happen in the tahoe node that is talking to the storage node and the client talking to the tahoe client node running on the local host would probably reply with a `504` (Gateway timed out) perhaps?